onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -984,6 +990,39 @@ of [N, 0] then [N, 0].
         updateOutputShape(ctx, 0, output_shape);
       });
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(MatMulInteger16)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(R"DOC(
+Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html.
+ The production MUST never overflow. The accumulation may overflow if and only if in 32 bits.)DOC")
+      .Input(0, "A", "N-dimensional matrix A", "T1")
+      .Input(1, "B", "N-dimensional matrix B", "T2")
+      .Output(0, "Y", "Matrix multiply results from A * B", "T3")
+      .TypeConstraint("T1", {"tensor(int16)", "tensor(uint16)"}, "Constrain input A data types as 8-bit integer tensor")
+      .TypeConstraint("T2", {"tensor(int16)", "tensor(uint16)"}, "Constrain input B data types as 8-bit integer tensor")
+      .TypeConstraint("T3",
+                      {"tensor(int32)", "tensor(uint32)"},
+                      "Constrain output Y data types as 32-bit integer tensor."
+                      "T3 must be tensor(uint32) when both T1 and T2 are tensor(uint16),"
+                      "or must be tensor(int32) when either T1 or T2 is tensor(int16).")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
```

Comment:
I wouldn't have expected something with a name of MatMulInteger16 to output int32 data. Should the output size be in the op name?

Reply:
This is to differentiate from MulMulInteger which takes int8/uint8 and outputs int32. It's in another PR #1391.

---
In reply to: [312416854](https://github.com/microsoft/onnxruntime/pull/881#discussion_r312416854) [](ancestors = 312416854)